test/query_snapshot_stream_manager_test.dart

```diff
@@ -2,37 +2,125 @@ import 'dart:async';
 import 'package:test/test.dart';
 
+class User {
+  final String name;
+
+  User(this.name);
+
+  User.fromMap(Map<String, dynamic> map) : name = map['name'];
+
+  Map<String, dynamic> toFirestore() => {
+        'name': name,
+      };
+}
+
 void main() {
   late FakeFirebaseFirestore instance;
 
   setUp(() async {
     instance = FakeFirebaseFirestore();
-    await instance.collection('users').add({
-      'name': 'Bob',
-    });
-    await instance.collection('users').add({
-      'name': 'Marie',
-    });
   });
 
-  test('Should allow concurrent cache modifications', () {
-    final subscriptions = <StreamSubscription<QuerySnapshot>>[];
-    instance
-        .collection('users')
-        .snapshots()
-        .listen((QuerySnapshot<Map<String, dynamic>> snapshot) {
-      // Add a new listener to the same query while an update being fired. This
-      // used to trigger a concurrency error.
-      subscriptions.add(instance
+  group('cache modifications', () {
+    setUp(() async {
+      await instance.collection('users').add({
+        'name': 'Bob',
+      });
+      await instance.collection('users').add({
+        'name': 'Marie',
+      });
+    });
+
+    test('Should allow concurrent cache modifications', () {
+      final subscriptions = <StreamSubscription<QuerySnapshot>>[];
+      instance
           .collection('users')
           .snapshots()
-          .listen((QuerySnapshot<Map<String, dynamic>> snapshot) {}));
+          .listen((QuerySnapshot<Map<String, dynamic>> snapshot) {
+        // Add a new listener to the same query while an update being fired. This
+        // used to trigger a concurrency error.
+        subscriptions.add(instance
+            .collection('users')
+            .snapshots()
+            .listen((QuerySnapshot<Map<String, dynamic>> snapshot) {}));
+      });
+      // Fire a snapshot update.
+      instance.collection('users').add({
+        'name': 'Steve',
+      });
+    });
+  });
+
+  group('listener updates', () {
+    late CollectionReference collectionRef;
+    setUp(() {
+      collectionRef = instance.collection('users');
     });
-    // Fire a snapshot update.
-    instance.collection('users').add({
-      'name': 'Steve',
+
+    void validate() {
+      expect(
+        collectionRef.snapshots(),
+        emitsInOrder(
+          [
+            isA<MockQuerySnapshot<Map<String, dynamic>>>().having(
+                (event) => (event.docs.first.data() as Map)['name'],
+                'has correct name',
+                'Bob'),
+            isA<MockQuerySnapshot<Map<String, dynamic>>>().having(
+                (event) => (event.docs.first.data() as Map)['name'],
+                'has correct name',
+                'Marie'),
+          ],
+        ),
+      );
+    }
+
+    test('can receive updates', () async {
+      await collectionRef.doc('1').set({
+        'name': 'Bob',
+      });
+
+      validate();
+
+      await pumpEventQueue();
```

Comment:
I tried removing all the `await pumpEventQueue()` and the tests still passed. Why are they needed?

Reply:
they're use of habit, I can remove them